### PR TITLE
AB#411718-Feature-Create export data for subsidiaries for a Compliance scheme

### PR DESCRIPTION
**Pull request by deorab**
… and extract the system generated subsidiary Ids for all child entities they provide data for

Assumption:
We are just returning the data, not creating a CSV We need someone to provide the file format

Note:
Selection needs to check the RelationToDate on OrganisationRelationships - any rows that have a non-null date that is <= now should not be returnd.

### CODE REVIEW

File: src/FacadeAccountCreation/FacadeAccountCreation.Core/Services/Organisation/OrganisationService.cs [hunk 1]
```diff
@@ -192,4 +193,27 @@ public async Task<OrganisationRelationshipModel> GetOrganisationRelationshipsByO
             _httpClient.DefaultRequestHeaders.Clear();
         }
     }
+
+    public async Task<List<ExportOrganisationSubsidiariesResponseModel>> ExportOrganisationSubsidiaries(Guid organisationExternalId)
+    {
+        HttpResponseMessage result = null;
+        var endpoint = $"{OrganisationGetSubsidiaryUri}/{organisationExternalId}/ExportOrganisationSubsidiaries";
```
Comment (mikeymonster) [inline src/FacadeAccountCreation/FacadeAccountCreation.Core/Services/Organisation/OrganisationService.cs:200]: Endpoint should be export-subsidiaries

File: src/FacadeAccountCreation/FacadeAccountCreation.Core/Services/Organisation/OrganisationService.cs [hunk 2]
```diff
@@ -192,4 +193,27 @@ public async Task<OrganisationRelationshipModel> GetOrganisationRelationshipsByO
...
+        var response = await _httpClient.GetAsync(endpoint);
+        try
+        {
+            _logger.LogInformation("Attempting to Export the Organisation Relationships for Organisation Id : '{organisationId}'", organisationExternalId);
```
Comment (mikeymonster) [inline src/FacadeAccountCreation/FacadeAccountCreation.Core/Services/Organisation/OrganisationService.cs:204]: Capital O inside {}

File: src/FacadeAccountCreation/FacadeAccountCreation.Core/Services/Organisation/OrganisationService.cs [hunk 3]
```diff
@@ -192,4 +193,27 @@ public async Task<OrganisationRelationshipModel> GetOrganisationRelationshipsByO
...
+            result = await _httpClient.GetAsync(endpoint);
+        }
+        catch (Exception e)
+        {
+            _logger.LogError(e, "Failed to Export Organisation Relationships for Organisation Id: '{organisationId}'", organisationExternalId);
```
Comment (mikeymonster) [inline src/FacadeAccountCreation/FacadeAccountCreation.Core/Services/Organisation/OrganisationService.cs:209]: Capital O inside {}

File: src/FacadeAccountCreation/FacadeAccountCreation.Core/Services/Organisation/OrganisationService.cs [hunk 4]
```diff
@@ -192,4 +193,27 @@ public async Task<OrganisationRelationshipModel> GetOrganisationRelationshipsByO
             _httpClient.DefaultRequestHeaders.Clear();
         }
     }
+
+    public async Task<List<ExportOrganisationSubsidiariesResponseModel>> ExportOrganisationSubsidiaries(Guid organisationExternalId)
+    {
+        HttpResponseMessage result = null;
+        var endpoint = $"{OrganisationGetSubsidiaryUri}/{organisationExternalId}/ExportOrganisationSubsidiaries";
+        var response = await _httpClient.GetAsync(endpoint);
```
Comment (mikeymonster) [inline src/FacadeAccountCreation/FacadeAccountCreation.Core/Services/Organisation/OrganisationService.cs:201]: You are calling this twice - once before and once inside the try

File: src/FacadeAccountCreation/FacadeAccountCreation.Core/Services/ComplianceScheme/ComplianceSchemeService.cs [hunk 5]
```diff
@@ -293,4 +296,30 @@ public async Task<HttpResponseMessage> GetAllReasonsForRemovalsAsync()
...
+
+    public async Task<List<ExportOrganisationSubsidiariesResponseModel>> ExportComplianceSchemeSubsidiaries(Guid userId, Guid organisationId, Guid complianceSchemeId)
+    {
+        HttpResponseMessage result = null;
+        var endpointConfigValue = $"{_config.GetSection("ComplianceSchemeEndpoints").GetSection("ExportComplianceSchemeSubsidiaries").Value}";
+        var endpoint = string.Format(endpointConfigValue, organisationId, complianceSchemeId);
+        var response = await _httpClient.GetAsync(endpoint);
+        try
+        {
+            _logger.LogInformation("Attempting to Export the Compliance Scheme Subsidiaries for Organisation Id : '{organisationId}'", organisationId);
```
Comment (mikeymonster) [inline src/FacadeAccountCreation/FacadeAccountCreation.Core/Services/ComplianceScheme/ComplianceSchemeService.cs:308]: Big O inside {}

File: src/FacadeAccountCreation/FacadeAccountCreation.Core/Services/ComplianceScheme/ComplianceSchemeService.cs [hunk 6]
```diff
@@ -293,4 +296,30 @@ public async Task<HttpResponseMessage> GetAllReasonsForRemovalsAsync()
...
+            _httpClient.DefaultRequestHeaders.Clear();
+            _httpClient.DefaultRequestHeaders.Add(XEprUserHeader, userId.ToString());
+            result = await _httpClient.GetAsync(endpoint);
+        }
+        catch (Exception e)
+        {
+            _logger.LogError(e, "Failed to Export the Compliance Scheme Subsidiaries for Organisation Id : '{organisationId}'", organisationId);
```
Comment (mikeymonster) [inline src/FacadeAccountCreation/FacadeAccountCreation.Core/Services/ComplianceScheme/ComplianceSchemeService.cs:315]: Big O inside {}

File: src/FacadeAccountCreation/FacadeAccountCreation.Core/Services/ComplianceScheme/ComplianceSchemeService.cs [hunk 7]
```diff
@@ -293,4 +296,30 @@ public async Task<HttpResponseMessage> GetAllReasonsForRemovalsAsync()
             throw;
         }
     }
+
+    public async Task<List<ExportOrganisationSubsidiariesResponseModel>> ExportComplianceSchemeSubsidiaries(Guid userId, Guid organisationId, Guid complianceSchemeId)
+    {
+        HttpResponseMessage result = null;
+        var endpointConfigValue = $"{_config.GetSection("ComplianceSchemeEndpoints").GetSection("ExportComplianceSchemeSubsidiaries").Value}";
+        var endpoint = string.Format(endpointConfigValue, organisationId, complianceSchemeId);
+        var response = await _httpClient.GetAsync(endpoint);
```
Comment (mikeymonster) [inline src/FacadeAccountCreation/FacadeAccountCreation.Core/Services/ComplianceScheme/ComplianceSchemeService.cs:305]: You are calling GetAsync twice

File: src/FacadeAccountCreation/FacadeAccountCreation.API/Controllers/ComplianceSchemesController.cs [hunk 8]
```diff
@@ -452,4 +453,26 @@ public async Task<IActionResult> RemoveComplianceSchemeMember(
             return HandleError.Handle(e);
         }
     }
+
+    [HttpGet]
+    [Route("{organisationId:guid}/schemes/{complianceSchemeId:guid}/ExportComplianceSchemeSubsidiaries")]
```
Comment (mikeymonster) [inline src/FacadeAccountCreation/FacadeAccountCreation.API/Controllers/ComplianceSchemesController.cs:458]: This should be export-subsidiaries

File: src/FacadeAccountCreation/FacadeAccountCreation.Core/Services/ComplianceScheme/ComplianceSchemeService.cs [hunk 9]
```diff
@@ -293,4 +296,30 @@ public async Task<HttpResponseMessage> GetAllReasonsForRemovalsAsync()
             throw;
         }
     }
+
+    public async Task<List<ExportOrganisationSubsidiariesResponseModel>> ExportComplianceSchemeSubsidiaries(Guid userId, Guid organisationId, Guid complianceSchemeId)
+    {
+        HttpResponseMessage result = null;
+        var endpointConfigValue = $"{_config.GetSection("ComplianceSchemeEndpoints").GetSection("ExportComplianceSchemeSubsidiaries").Value}";
```
Comment (mikeymonster) [inline src/FacadeAccountCreation/FacadeAccountCreation.Core/Services/ComplianceScheme/ComplianceSchemeService.cs:303]: The config needs to be added to app.config

    "ExportComplianceSchemeSubsidiaries": "api/compliance-schemes/{0}/schemes/{1}/export-subsidiaries"

**Review by RamakrishnaGujjula — APPROVED** [review]

**Review by tsnatos — APPROVED** [review]

**Review by mikeymonster — APPROVED** [review]

**Pull request merged**